ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite_core.cc

```diff
@@ -437,68 +384,311 @@ TFLiteCore::setInputTensorDim (const GstTensorsInfo * info)
-    tensor_ptr->data.raw = (char *) output[i].data;
-    tensors_idx.push_back (tensor_idx);
+  if (accelerators == NULL) {
+    goto use_nnapi_ini;
   }
 
-  for (int i = 0; i < inputTensorMeta.num_tensors; ++i) {
-    tensor_idx = interpreter->inputs ()[i];
-    tensor_ptr = interpreter->tensor (tensor_idx);
+  /* If set by user, get the precise accelerator */
+  use_nnapi = (bool) g_regex_match_simple (REGEX_ACCL_NNAPI, accelerators,
+      G_REGEX_CASELESS, G_REGEX_MATCH_NOTEMPTY);
+  if (use_nnapi == TRUE) {
+    /** Default to auto mode */
+    accelerator = ACCL_AUTO;
+    nnapi_elem = g_regex_new (REGEX_ACCL_NNAPI_ELEM, G_REGEX_CASELESS,
+        G_REGEX_MATCH_NOTEMPTY, NULL);
 
-    g_assert (tensor_ptr->bytes == input[i].size);
-    tensor_ptr->data.raw = (char *) input[i].data;
-    tensors_idx.push_back (tensor_idx);
+    /** Now match each provided element and get specific accelerator */
+    if (g_regex_match (nnapi_elem, accelerators, G_REGEX_MATCH_NOTEMPTY,
+          &match_info)) {
+
+      while (g_match_info_matches (match_info)) {
+        gchar *word = g_match_info_fetch (match_info, 0);
+        accelerator = get_accl_hw_type (word);
+        g_free (word);
+        break;
+      }
+    }
+    g_match_info_free (match_info);
+    g_regex_unref (nnapi_elem);
+  } else  {
+    goto use_nnapi_ini;
   }
 
-#ifdef ENABLE_TFLITE_NNAPI_DELEGATE
-  if (use_nnapi)
-    status = nnfw_delegate->Invoke (interpreter.get());
-  else
-#endif
-    status = interpreter->Invoke ();
+  return;
 
-  /** if it is not `nullptr`, tensorflow makes `free()` the memory itself. */
-  int tensorSize = tensors_idx.size ();
-  for (int i = 0; i < tensorSize; ++i) {
-    interpreter->tensor (tensors_idx[i])->data.raw = nullptr;
+use_nnapi_ini:
+  use_nnapi = nnsconf_get_custom_value_bool ("tensorflowlite", "enable_nnapi",
+      FALSE);
+  if (use_nnapi == FALSE) {
+    accelerator = ACCL_NONE;
+  } else {
+    accelerator = ACCL_AUTO;
   }
+}
 
-#if (DBG)
-  gint64 stop_time = g_get_real_time ();
-  g_message ("Invoke() is finished: %" G_GINT64_FORMAT,
-      (stop_time - start_time));
-#endif
-
-  if (status != kTfLiteOk) {
-    g_critical ("Failed to invoke");
+/**
+ * @brief	initialize the object with tflite model
+ * @return 0 if OK. non-zero if error.
+ *        -1 if the model is not loaded.
+ *        -2 if the initialization of input tensor is failed.
+ *        -3 if the initialization of output tensor is failed.
+ */
+int
+TFLiteCore::init ()
+{
+  if (loadModel ()) {
+    g_critical ("Failed to load model\n");
     return -1;
   }
+  if (setInputTensorProp ()) {
+    g_critical ("Failed to initialize input tensor\n");
+    return -2;
+  }
+  if (setOutputTensorProp ()) {
+    g_critical ("Failed to initialize output tensor\n");
+    return -3;
+  }
+  return 0;
+}
+
+/**
+ * @brief	get the model path
+ * @return the model path.
+ */
+const char *
+TFLiteCore::getModelPath ()
+{
+  return model_path;
```

Comment:
add lock here, as reload model can update this within its lock?

Reply:
Updated.

Reply:
I think this function should return a copy of the model_path stored inside interpreter, as reloadModel can free this model_path making the return value of this function a hazard.

Reply:
True. But, I just changed `getModelPath()` to `compareModelPath()` because it's used only to compare model path in `tensor_filter_tensorflow_lite.c`. Making/freeing its duplicate seems not efficient.

Reply:
compareModelPath is good.